app/models/product_importer.rb

```diff
@@ -0,0 +1,314 @@
+
+class ProductImporter
+  extend ActiveModel::Naming
+  include ActiveModel::Conversion
+  include ActiveModel::Validations
+
+  def initialize(file, editable_enterprises, options={})
+    if file.is_a?(File)
+      @file = file
+      @options = options
+      @sheet = open_spreadsheet
+      @valid_entries = {}
+      @invalid_entries = {}
+
+      @products_to_create = []
+      @variants_to_create = []
+      @variants_to_update = []
+      
+      @products_created = 0
+      @variants_created = 0
+      @variants_updated = 0
+
+      @editable_enterprises = {}
+      editable_enterprises.map { |e| @editable_enterprises[e.name] = e.id }
+
+      validate_all
+    else
+      self.errors.add(:importer, "error: no file uploaded")
+    end
+  end
+
+  def persisted?
+    false #ActiveModel, not ActiveRecord
+  end
+
+  # Private methods below which only work with a valid spreadsheet object can be called publicly
+  # via here if the spreadsheet was successfully loaded, otherwise they return nil (without error).
+  def method_missing(method, *args, &block)
+    if self.respond_to?(method, include_private=true)
+      @sheet ? self.send(method, *args, &block) : nil
+    else
+      super
+    end
+  end
+
+  private
+
+  def open_spreadsheet
+    if accepted_mimetype
+      @sheet = Roo::Spreadsheet.open(@file, extension: accepted_mimetype)
+    else
+      self.errors.add(:importer, "could not proccess file: invalid filetype")
+      nil
+    end
+  end
+
+  def accepted_mimetype
+    File.extname(@file.path).in?(['.csv', '.xls', '.xlsx', '.ods']) ? @file.path.split('.').last.to_sym : false
+
+    # case @file.content_type
+    #   when "text/csv"
+    #     :csv
+    #   when "application/excel", "application/x-excel", "application/x-msexcel", "application/vnd.ms-excel"
+    #     :xls
+    #   when "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
+    #     :xlsx
+    #   when "application/vnd.oasis.opendocument.spreadsheet"
+    #     :ods
+    #   else
+    #     #Mimetype not compatible
+    #     false
+    # end
+  end
+
+  def headers
+    @sheet.row(1)
+  end
+
+  def rows
+    (2..@sheet.last_row).map do |i|
+      @sheet.row(i)
+    end
+  end
+
+  def entries
+    rows.map do |row|
+      Hash[[headers, row].transpose]
+    end
+  end
+
+  def validate_all
+    suppliers_index = @suppliers_index || get_suppliers_index
+    categories_index = @categories_index || get_categories_index
+
+    entries.each_with_index do |entry, i|
+
+      # Fetch/assign and validate supplier id
+      supplier_name = entry['supplier']
+      if supplier_name.blank?
+        invalidate({i+2 => {entry: entry, errors: ["Supplier name field is empty"]}}) #unless entry['supplier_id']
+      else
+        if suppliers_index[supplier_name]
+          entry['supplier_id'] = suppliers_index[supplier_name]
+        else
+          invalidate({i+2 => {entry: entry, errors: ["Supplier \"#{supplier_name}\" not found in database"]}})
```

Comment:
Yeah this is something I thought needed tidying up. It's because Roo counts the rows starting from 1, but the first line of actual data is on line 2 (because of the headers), whereas Ruby obviously counts array items from 0.